flocker/node/agents/test/test_blockdevice.py

```diff
@@ -5249,3 +5257,231 @@ def test_get_device_path_until_detach(self):
 
         self.assertRaises(UnattachedVolume,
                           self.cache.get_device_path, attached_id1)
+
+
+@attributes(["_proxy_object", "_callback"], apply_with_init=False)
+class CallbackProxy(object):
+    """
+    A proxy to another object that calls a callback before and after each call
+    to the underlying object. This enables tests where you want to inject code
+    before or after any calls to an underlying API.
+
+    Note that this proxy assumes that the underlying object only has public
+    methods and not public attributes.
```

Comment:
So I'll probably do this cause that code looks interesting, but I'm not 100% it adds that much, and it comes at some complexity.

@tomprince 
What's the value add for this? Another layer of double checking the consumer of this interface is only using the interface and not cheating by using non-interface methods and variables?

Feel free to respond out of band as I do plan on using the interface_decorator for this.